src/include/rbd/librbd.hpp

```diff
@@ -35,9 +39,51 @@ namespace librbd {
   typedef void (*callback_t)(completion_t cb, void *arg);
 
   typedef struct {
+    void dump(ceph::Formatter *f) const {
+	f->dump_string("name", "user");
+    }
+
+    std::string to_string() const {
+      return "user snapshot";
+    }
+  } user_snap_namespace_t;
+
+  typedef struct {
+    void dump(ceph::Formatter *f) const {
+	f->dump_string("name", "unknown");
+    }
+
+    std::string to_string() const {
+      return "unknown snapshot";
+    }
+  } unknown_snap_namespace_t;
+
+  typedef struct {
+    std::string group_pool;
+    std::string group_name;
+    std::string snap_name;
+
+    void dump(ceph::Formatter *f) const {
+	f->dump_string("name", "group");
+	f->dump_string("group_pool", group_pool);
+	f->dump_string("group_name", group_name);
+	f->dump_string("group_snap_name", snap_name);
+    }
+
+    std::string to_string() const {
+      return "group snapshot member - " + group_pool +
+				   "/" + group_name +
+				   "@" + snap_name;
+    }
+  } group_snap_namespace_t;
+
+  typedef boost::variant<user_snap_namespace_t, group_snap_namespace_t, unknown_snap_namespace_t> snap_namespace_t;
+
+  typedef struct {
     uint64_t id;
     uint64_t size;
     std::string name;
+    snap_namespace_t snap_namespace;
```

Comment:
Since you will have to support the C API as well, and you might as well keep the interfaces semi-similar, perhaps add a new snapshot namespace enum, a new ```snap_get_namespace_type``` call that will retrieve the namespace enum, and a new ```snap_get_group_spec``` (or similar) method to retrieve the group information should the namespace type by a group.